packages/web3-providers/src/Approval/approvalListState.ts

```diff
@@ -0,0 +1,98 @@
+import { produce } from 'immer'
+import { type ChainId } from '@masknet/web3-shared-evm'
+import type { BigNumber } from 'bignumber.js'
+
+export type TokenApprovalInfoAccountMap = Record<
+    string,
+    {
+        [key in ChainId]?: {
+            spenderList: Record<string, Record<string, { amount: BigNumber; transactionBlockNumber: number }>>
+            fromBlock: number
+        }
+    }
+>
+
+export type NFTApprovalInfoAccountMap = Record<
+    string,
+    {
+        [key in ChainId]?: {
+            spenderList: Record<string, Record<string, { approved: boolean; transactionBlockNumber: number }>>
+            fromBlock: number
+        }
+    }
+>
+
+export class ApprovalListState {
+    private _token_state: TokenApprovalInfoAccountMap = {}
+    private _nft_state: NFTApprovalInfoAccountMap = {}
+
+    public get tokenState() {
+        return this._token_state
+    }
+
+    public get nftState() {
+        return this._nft_state
+    }
+
+    public updateTokenState(
+        account: string,
+        spender: string,
+        address: string,
+        chainId: ChainId,
+        fromBlock: number,
+        transactionBlockNumber: number,
+        amount: BigNumber,
+    ) {
+        this._token_state = produce(this._token_state, (draft) => {
+            if (!draft[account]) draft[account] = {}
```

Comment:
If we are using EVM addresses as key, they need to be formatted.

Reply:
done